Blog post: May 2022 Security Announcement (#928)
fluxcd/website: content/en/blog/2022-05-10-may-2022-security-announcement.md

---
author: 'dholbach & pjbgf'
date: 2022-05-10 8:30:00+00:00
title: May 2022 Security Announcement

> description: New Flux and Flagger releases bring you the latest and greatest just yet! KubeCon news, GitOps Days coming up and lots of new resources - potentially the most newly added videos in particular! Lots of great contributions from our community. 💖

Reviewer: I was thinking we could use the TL;DR... 

```suggestion
description: The Flux Team has found three security vulnerabilities in Flux, Today we will go through them and talk about what this may mean to you. We strongly advise you to upgrade your clusters as soon as you can. :lock: 
```

| CVE-2022-24877 | [Improper path handling in Kustomization files allows path traversal](https://github.com/fluxcd/flux2/security/advisories/GHSA-j77r-2fxf-5jrw) | Critical | `< v0.29.0`
| CVE-2022-24878 | [Improper path handling in Kustomization files allows for denial of service](https://github.com/fluxcd/flux2/security/advisories/GHSA-7pwf-jg34-hxwp) | High | `< v0.29.0 >= v0.19.0`

Breaking changes to be aware of in the upgrade process:
[0.29](/blog/2022/05/april-2022-update/#latest-flux-release-series-is-029),
[0.28](https://github.com/fluxcd/flux2/discussions/2567),
[0.27](/blog/2022/03/february-update/#latest-flux-is-027),
[0.26](/blog/2022/01/january-update/#flux-v026-more-secure-by-default),
[0.24 - 0.21](/blog/2021/11/december-update/#a-flurry-of-flux-releases).

If you cannot immediately update or are hard pressed for time and need a
work-around for now, please see the CVE advisories linked above for more
information.

## Some Background

Last week the Flux Security Team disclosed three new vulnerabilities

> which affect most versions below v0.29, and have a greater impact in

Reviewer 2: ```suggestion
which affect v0.28 and older versions, and have a greater impact in
```

multi-tenancy deployments.

The reason why the impact is greater in multi-tenancy deployments is due
to the way that Flux/GitOps works. Flux tends to operate like a cluster
admin, having permissions to apply any changes to a cluster, regardless
of their scope - at namespace or cluster levels. Users having access to
a source repository, or simply having access to create/alter Flux
objects within a cluster can instruct Flux to apply such changes, which
in single tenancy effectively means that such users have cluster admin
permissions to the target clusters. The caveat being that using Flux you
can add additional security controls between the user and the target
cluster, for example, before each change is merged into a repository
Pull Requests must be created requiring peer reviews. The Open GitOps
community started defining and codifying this further - have a look [at
this blog post](https://opengitops.dev/blog/sec-gitops/) if
you want to know more.

In multi-tenancy environments, users with similar permissions can only
affect part of a cluster, or an isolated cluster in a group of clusters.
Therefore, if a user can privilege escalate (or deny service) at a Flux

> level, it will have an impact way larger than on single tenancies, as

Reviewer 2: ```suggestion
level, it will have an impact way larger than on single-tenant clusters, as
```

> level, it will have an impact way larger than on single tenancies, as
> the users in a way already had such permissions - maybe with some
> balances and checks.

Reviewer: Was thinking on potentially rephrasing this. Keen on other opinions:
```suggestion
level, it will have an impact way larger than on single tenancies, as
the users can impact more than just themselves.
```

At time of writing, we are unaware of any public exploits in the wild,
and therefore have no reason to believe such vulnerabilities have been
actively exploited.

## The advisories in detail

### [CVE-2022-24817](https://github.com/fluxcd/flux2/security/advisories/GHSA-vvmq-fwmg-2gjc) - Kubeconfig Validation

> At the beginning of the Flux2 journey we implemented a feature to apply

Reviewer: ```suggestion
At the beginning of the Flux2's journey we implemented a feature to apply
```

state to remote clusters. It enables users to have a management cluster
in which Flux is installed, which then applies changes to other
clusters, making it ideal for some multi-tenancy scenarios in which high
isolation across tenants is needed.

The connection between the management cluster and the target clusters is

> done via a kubeconfig which is set at the `spec.KubeConfig` field in

Reviewer 2: ```suggestion
done by referencing a Kubernetes secret containing a kubeconfig, which is set at the `spec.KubeConfig` field in
```

either `Kustomization` or `HelmRelease` objects.

One interesting thing about kubeconfigs is that they are quite
extensible. You can for example define an executable which will then be
automatically called by `kubectl` every time it requires a new on-demand
access token. An example of this in action is
[aws-iam-authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator),

> which enables AWS users to authenticate against AWS and reuse JWT tokens
> to access EKS clusters seamlessly.

Reviewer: Not sure the initial attempt was clear enough to folks that are not aware of its ins and outs:
```suggestion
which enables AWS users to authenticate against AWS and then use the
returning JWT tokens to access their EKS clusters. Once the token expires,
the process happens again. All that is managed by `kubectl` behind the scenes
without user intervention.
```

of AppArmor and SELinux profiles to enforce at Kernel level what
binaries could be executed.

### [CVE-2022-24877](https://github.com/fluxcd/flux2/security/advisories/GHSA-j77r-2fxf-5jrw) - Kustomization Path Traversal

Flux allows users to lean on Kustomize features to make their lives
easier as they go on about declaring the state of their clusters. Some
of those features could result in sensitive data from the pod filesystem
to be exposed into the target cluster, which could lead to a malicious
tenant being privy to anything sensitive that may exist in the
controller's filesystem or attached volumes (e.g. token).

The mitigation for the path traversal was to create stronger bounds,
enforcing that all `kustomization` operations happen within such bounds or
result in error.

### [CVE-2022-24878](https://github.com/fluxcd/flux2/security/advisories/GHSA-7pwf-jg34-hxwp) - Kustomization Denial of Service

Whilst working on the mitigation of the previous CVE, we have noticed
that in some scenarios a specially crafted `kustomization.yaml` could lead
to the `kustomize-controller` to enter into an endless loop, and finally
crash.

> In single tenancy, this would mean that an user may make a mistake and

Reviewer 2: ```suggestion
For single-tenant clusters, this would mean that an user may make a mistake and
```

the controller stops working, potentially resulting in future

> reconciliations not being applied. In multi-tenancy, depending on the

Reviewer 2: ```suggestion
reconciliations not being applied. For multi-tenant clusters, depending on the
```

> deployment model, tenant A could cause the disruption to affect not only

Reviewer 2: ```suggestion
deployment model, a tenant could cause a disruption to affect not only
```

itself but also the other tenants and potentially even the management
cluster.

The solution mitigating this vulnerability was to further improve our
validation and ensure that such scenarios are not processed in the first
place.

## Inspecting your Flux system version

> To check if your system is currently vulnerable, run `flux version`
> with your `KubeConfig` set to the cluster you want to inspect. This will

Reviewer 2: ```suggestion
To check if your system is currently vulnerable, run `flux version --context=my-cluster`
with the `--context` set to the cluster you want to inspect. This will
```

:collision: If you find the controllers versioned within the range
mentioned above, follow the upgrade procedure for your system. For
`flux bootstrap`, this can be done by running the command again with
the same arguments as used during install.

:warning: Please note that if you are upgrading from below one of the
versions in the following list, there are breaking changes and, pre-
and/or post-upgrade notes you need to take into account:
[0.29](/blog/2022/05/april-2022-update/#latest-flux-release-series-is-029),
[0.28](https://github.com/fluxcd/flux2/discussions/2567),
[0.27](/blog/2022/03/february-update/#latest-flux-is-027),
[0.26](/blog/2022/01/january-update/#flux-v026-more-secure-by-default),
[0.24 - 0.21](/blog/2021/11/december-update/#a-flurry-of-flux-releases).

### Up-to-date Flux system

An up to date Flux system should at least have versions listed below:

- flux `>= v0.29.0`
- helm-controller `>= v0.19.0`
- kustomize-controller `>= v0.24.0`

So in practice the output could look like this:

```cli
$ flux version
flux version
flux: v0.30.2
helm-controller: v0.21.0
kustomize-controller: v0.25.0
notification-controller: v0.23.5
source-controller: v0.24.4
```

> 

Reviewer 2: ```suggestion

We encourage all users to keep Flux up-to-date. We offer a [GitHub Action](https://github.com/fluxcd/flux2/tree/main/action#automate-flux-updates) with which you can automate the Flux upgrades in a GitOps manner, without having to connect from CI to the cluster's API, as Flux is capable of upgrading itself from Git.
```
